validator/proposer/service.go

```diff
@@ -159,43 +160,93 @@ func (p *Proposer) run(done <-chan struct{}, client pb.ProposerServiceClient) {
-			p.lock.Lock()
-
-			bitmask := p.GenerateBitmask(p.pendingAttestation)
-			// TODO(#619): Implement real proposals with randao reveals and attestation fields.
-			req := &pb.ProposeRequest{
-				ParentHash: latestBlockHash[:],
-				// TODO(#511): Fix to be the actual, timebased slot number instead.
-				SlotNumber:         latestBeaconBlock.GetSlot() + 1,
-				RandaoRevealHash32: []byte{},
-				AttestationBitmask: bitmask,
-				Timestamp:          ptypes.TimestampNow(),
-			}
-			res, err := client.ProposeBlock(p.ctx, req)
-			if err != nil {
-				log.Errorf("Could not propose block: %v", err)
-				continue
-			}
+func (p *Proposer) receiveAssignment(latestBeaconBlock *pbp2p.BeaconBlock, client pb.ProposerServiceClient) {
+	log.Info("Performing proposer responsibility")
 
-			log.Infof("Block proposed successfully with hash %#x", res.BlockHash)
-			p.pendingAttestation = nil
-			p.lock.Unlock()
-		}
+	block, err := p.computeBlockToBeProposed(latestBeaconBlock, client)
+	if err != nil {
+		log.Error(err)
+		return
 	}
+
+	proposalData, err := p.createProposalDataFromBlock(block)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	signature := p.signProposalData(proposalData)
+	block.Signature[0] = signature[:]
+
+	hash, err := client.ProposeBlock(p.ctx, block)
+	if err != nil {
+		log.Errorf("Unable to propose block %v", err)
+		return
+	}
+
+	log.Infof("Successfully proposed block with hash %#x", hash.BlockHash)
+}
+
+func (p *Proposer) computeBlockToBeProposed(latestBlock *pbp2p.BeaconBlock,
+	client pb.ProposerServiceClient) (*pbp2p.BeaconBlock, error) {
+	// Extract the hash of the latest beacon block to use as parent hash in
+	// the proposal.
+	if latestBlock == nil {
+		return nil, fmt.Errorf("Could not marshal nil latest beacon block")
+
+	}
+	data, err := proto.Marshal(latestBlock)
+	if err != nil {
+		return nil, fmt.Errorf("Could not marshal latest beacon block: %v", err)
+	}
+	latestBlockHash := hashutil.Hash(data)
+
+	// To prevent any unaccounted attestations from being added.
+	p.lock.Lock()
+
+	bitmask := p.GenerateBitmask(p.pendingAttestation)
+	// TODO(#619): Implement real proposals with randao reveals and attestation fields.
+	req := &pb.ProposeRequest{
+		ParentHash: latestBlockHash[:],
+		// TODO(#511): Fix to be the actual, timebased slot number instead.
+		SlotNumber:         latestBlock.GetSlot() + 1,
+		RandaoRevealHash32: []byte{},
+		AttestationBitmask: bitmask,
+		Timestamp:          ptypes.TimestampNow(),
+	}
+	res, err := client.ComputeBlockWithStateRoot(p.ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("Could not compute new block: %v", err)
+	}
+
+	log.Infof("Block has state root hash of %#x", res.StateRootHash32)
+	p.pendingAttestation = nil
+	p.lock.Unlock()
+
+	return res, nil
+}
+
+func (p *Proposer) createProposalDataFromBlock(block *pbp2p.BeaconBlock) (*pbp2p.ProposalSignedData, error) {
+
+	marshalledBlock, err := proto.Marshal(block)
```

Comment:
I prefer to call these encodedBlock, but just something to keep in mind for naming consistency.